Add actual (previous) price sensor for amberelectric

upsuper opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

This is a revival of #119734, which has been approved by the integration owner @madpilot but got shot down by project maintainer @joostlek for mysterious reason and unclear path forward. I have to say I was very disappointed by @joostlek, and their comments made me so exasperated that I closed that PR impulsively.

I still think merging this change into the mainline is the best path forward. It would take more effort overall if I have to fork it just for this change.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This change adds sensors for actual prices for the last period, as well as past 24hrs into its attributes.

While past prices are recorded as history in Home Assistant, there is no guarantee that the actual price decided at the end of the period actually matches the current price last fetched. It is also not very easy to use a historical value in templates.

Adding the actual past prices as sensor enables creating dashboards similar to what Amber app shows to users which includes both past prices and forecast prices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

upsuper committed:
Include actual intervals in attributes of amberelectric price sensor

upsuper committed:
Merge branch 'dev' into amberelectric-actuals

upsuper committed:
put actual into new sensor

upsuper committed:
Merge branch 'dev' into amberelectric-actuals

upsuper committed:
Rename sensor to Previous

upsuper committed:
Fix sensor names in tests

upsuper committed:
Merge branch 'dev' into amberelectric-actuals

upsuper committed:
Update tests

joostlek commented:
Okay so the thing is, we don't want to put everything into attributes. The reason for that is that that fills up the database for every user, causing big state writes. So since we have service responses (or actions, however you call it), we want to move data objects that can't be poured into entities to the services. This way, the user can define what data they find relevant to store. Examples of this can be found [here](https://www.home-assistant.io/integrations/template/#trigger-based-handling-of-action-response-data).

And yes, it does add some complexity, but that's a problem we are facing as project as well. But we'd rather use this for new data, than merge this and have every user adapt to another way in a few months.

upsuper commented:
The problem with using services / actions is that they don't get updated automatically like state, and thus it's harder to build downstream in a data-driven / observable way. For example, in this case, some information may have changed without the state value updated. How would we solve this? Maybe adding a new state for the last update timestamp? But if we do that, it might be better that the whole integration be migrated to that first, rather than adding it just for this new one.

joostlek commented:
But how does the polling work currently?

upsuper commented:
Also I don't think integrations generally use services for providing information, but instead to do actions. This is a very important distinguish in my opinion. For example, a third-party dashboard may read state attributes to enhance the information they show, but one would not expect them to call a service to do so.

If big write to database or bloating it is the main concern, I think it might be better to introduce something like a "transient additional information" of state that never gets persisted, but still triggers a state update that observers can use, rather than saying you must use services for it.

upsuper commented:
> But how does the polling work currently?

My understanding is that polling is one of the method integrations use to generate value, and state is the abstraction integrations provide over whatever approach they use to generate such value.

Am I missing something? Is polling a thing that many services provide and dashboards use downstream?

joostlek commented:
Let me rephrase, how does the integration currently get its data. (I am now going to respond to the other comment haha)

upsuper commented:
> Let me rephrase, how does the integration currently get its data.

Sure, integration can poll upstream to get its data. But I think the idea is that by abstracting it as a value ignoring how it was imported, be it polling or hook callback, it is easier for downstream (like dashboard) to use.

joostlek commented:
> Also I don't think integrations generally use services for providing information, but instead to do actions

I don't really see what you mean with this, as in, services == actions and service responses == action responses. Not sure if that is what you are pointing at.

> For example, a third-party dashboard may read state attributes to enhance the information they show, but one would not expect them to call a service to do so.

Why not? The weather integrations and todo lists already do it like this. And you can still just create 1 entity and put everything in it, but then it's at least not by default

> I think it might be better to introduce something like a "transient additional information" of state that never gets persisted

I have no idea if this is thought about, and why/why not this is a good idea. So I can't say anything about that.

joostlek commented:
I think I misunderstood your problem here

> The problem with using services / actions is that they don't get updated automatically like state, and thus it's harder to build downstream in a data-driven / observable way. For example, in this case, some information may have changed without the state value updated.

I don't see how that is different, as you can just create a time interval to poll every x minutes to update the sensors you expose. Hence my question how the data is retrieved.

upsuper commented:
> I don't see how that is different, as you can just create a time interval to poll every x minutes to update the sensors you expose.

It's unclear what do you mean by "update the sensors" here.

If you mean updating Home Assistant sensor, then one may have to create tens of sensors to expose such information, and it doesn't really help with the bloating and big write issue since you still have this much information get poured into the database. (And I'm not sure whether there is an easy way to programmatically create a large set of sensors without writing a whole integration.) It also means downstream dashboard would suddenly have to start monitoring tens of sensors, each of them need a unique identifier, rather than just one and extract all the information from it.

If you mean having the downstream dashboard create its own structure to query, then the problem as I described above, it would no longer be a simple data-driven view, but would need to refactor into something able to track the asynchronousness of service calls.

Also if this is a common problem that integrations currently put a lot of things into attributes, forcing every single one of them to migrate to service call means a big breaking change for many downstream apps and they would have to do complicated refactor to adopt the new service structure. I'm not sure this is an appropriate path forward.

upsuper commented:
In addition, asking user to create an interval to poll data from integration has another problem: it could be out-of-sync. Because there is no mechanism for integration to tell the downstream sensor that the data is updated.

Attributes ensures that the data is updated atomically with state value itself, so that one can rely on that they are integrated. But requiring another level of polling means it becomes more error prone if downstream still relies on the integrity of the information.

upsuper commented:
> I have no idea if this is thought about, and why/why not this is a good idea. So I can't say anything about that.

It would be good to think about it. I have a strong feeling that requiring using service calls to fetch data in general is misguided and can increasingly be a problem. State is a good abstraction for observing data change, and it's better things continue to be built on top of it, and leave services for actual actions and possibly some more expensive data derivations.

joostlek commented:
> In addition, asking user to create an interval to poll data from integration has another problem: it could be out-of-sync. Because there is no mechanism for integration to tell the downstream sensor that the data is updated.

You can see when an entity is updated and run at the same time

> Also if this is a common problem that integrations currently put a lot of things into attributes, forcing every single one of them to migrate to service call means a big breaking change for many downstream apps and they would have to do complicated refactor to adopt the new structure using service to fetch data

We are not forcing them to change to attributes, but that's why we try to gatekeep new occurrences as much as possible. We do strongly recommend migrating to separate entities or actions. 

> It's unclear what do you mean by "update the sensors" here.

Receiving new state from the service to use in entities.

> and it doesn't really help with the bloating and big write issue since you still have this much information get poured into the database

We can store states themselves more efficiently and the user is in control of how much they store and what they want to use

upsuper commented:
> You can see when an entity is updated and run at the same time

Firstly, how one could know that an entity is updated when its value is not changed and it doesn't have such information in its attributes?

Secondly, even if one invokes the service at the same time the entity is updated, service call is still asynchronous, and a downstream would observe an invalid intermediate state before the service call is returned.

For example, let's say we have a entity say the current price of something, and we have a service call to get the previous price, and we have a derived state says whether the price increases or decreases. Assuming that the price changes once a minute, then the derived state would be updated twice each minute, the first updating it to no change, and the second updating it to the actual result. This isn't great.

> We are not forcing them to change to attributes, but that's why we try to gatekeep new occurrences as much as possible.

Without a good alternative, gatekeeping like this just makes it much harder to contribute.

upsuper commented:
> You can still poll the data with the coordinator so you have it available but instead of adding it all to attributes of the sensors let's make an action call so the user can retrieve that information as they prefer eg. use a trigger template.

Yes, but as I mentioned above, that would mean either
* downstream applications would be complicated as they can no longer be a simple observer of state, but has to handle async and polling, or
* if separate helper entity is created, they will not be updated synchronously with the main entity, breaking integrity of the data.

It would probably be more productive if you can point out what I'm missing above, rather than repeating what has already been discussed.

gjohansson-ST commented:
I have a hard time understanding why past prices are so critical as you refer them too.
You're right in terms of using a trigger template it would be updated after the main entity or whatever sequence the user prefers. 
But if that's such critical info that it needs to be available at all times why are they not sensors then instead of dumping it into attributes?

upsuper commented:
> I have a hard time understanding why past prices are so critical as you refer them too.

If you want a reason for this PR, that is because it is useful in dashboard to show the historical price.

This is an independent dashboard I built previously that shows the past, current, and predicted future price:
![image](https://github.com/user-attachments/assets/d0a3e6d6-828d-461d-bf95-ced1933551ad)

I'm trying to migrate this into Home Assistant as HA already has an integration for Amber and it would be good to have it serve as the single source of the information rather than having the dashboard fetching it separate.

> But if that's such critical info that it needs to be available at all times why are they not sensors then instead of dumping it into attributes?

I don't think it is appropriate to simply categorize things into critical and non-critical.

Adding more sensors is harder to manage, especially when the number is large. In this case, it is 48 values with some additional information like the time and descriptor of the value. So to support this dashboard with every single information present in a separate sensor, that is 48x3=144 sensors. And given the same argument applies equally to the future predicts (which already uses attribute), if we migrate that as well, that would need another 144 sensors. There are actually more useful information that such dashboard may make use of, such as whether the price is spiking, whether the given time period is in demand window. This is bloating very quickly.

These information, while useful for building a useful visualization, has very little value for user to look at directly, and thus exposing them as independent sensors just adds more distractions everywhere else.

upsuper commented:
They are "critical" just because so that we can have a dashboard with integrity of information, not that they are so critical that they have to bother users.

upsuper commented:
> So to support this dashboard with every single information present in a separate sensor, that is 48x3=144 sensors.

Also as I mentioned before, if you have all of them in separate sensors, you then need to give each sensor a unique identifier so that a dashboard may use them. This is much more cumbersome than if they are organized in a structured way associated with a single sensor.

madpilot commented:
A single actual sensor would always read the correct actual price, which fixes the history revision issue you have. Perhaps that’s a better compromise?

Call the sensor “Last actual”, and use the history for that plus the current price to build the chart?

it also means you don’t have to fetch the previous 48 intervals, you only need one.

upsuper commented:
> use the history for that plus the current price to build the chart?

It could work for this case, yes. It would still complicate the dashboard because it still needs to handle extra async operation rather than being a passive observer, but using history at least has a better semantics, and given it's a built-in functionality in Home Assistant, it is architecturally easier to handle than an action call defined by integration.

But one thing worries me is that the argument around not putting large data in attribute applies equally to the predictions, for which history wouldn't help.

In addition to that,
> A single actual sensor would always read the correct actual price

I'm not sure how true this is. I do notice that AEMO occasionally [reviews](https://aemo.com.au/market-notices?marketNoticeFacets=PRICES+SUBJECT+TO+REVIEW) the historical prices and [adjusts](https://aemo.com.au/market-notices?marketNoticeFacets=PRICE+ADJUSTMENT) them in some rare cases.

But to be honest, I don't know whether Amber's API handles this kind of cases at all, and I don't really care about it personally given its rarity and being unactionable.

madpilot commented:
> I'm not sure how true this is. I do notice that AEMO occasionally [reviews](https://aemo.com.au/market-notices?marketNoticeFacets=PRICES+SUBJECT+TO+REVIEW) the historical prices and [adjusts](https://aemo.com.au/market-notices?marketNoticeFacets=PRICE+ADJUSTMENT) them in some rare cases.

> But to be honest, I don't know whether Amber's API handles this kind of cases at all, and I don't really care about it personally given its rarity and being unactionable.

Technically this is true, and the Amber API does return the corrected price after the fact. But it happens so rarely that if you really care, you can update the history in HA manually.

gjohansson-ST commented:
I think @madpilot gave a good proposal so one sensor for actual price and one sensor for last actual price (so any backend correction would be seen in this one).
In addition to this any further historical information should be in an action call likewise any forecast should also be in an action call.

So in summary to get it complete, 2 sensors, 2 action calls.
It would be good to have the two sensors in this PR and then adding the custom action calls in two separate PR's once this is merged.

upsuper commented:
> any forecast should also be in an action call.

I would strongly be against this. Forecast is different from actual value. For actual values you could utilize the sensor history, but forecast you can't. I do need all the forecast values.

If the conclusion from Home Assistant team is that you absolutely want them to be in action call, and don't want to consider something like a transient non-persistent replacement of attributes for this kind of use cases, I guess the only option I have would really be to fork it for myself.

gjohansson-ST commented:
@upsuper it's the path forward for HA in general so we're not going to make anything different here.

But I don't understand why you would think this is such a drawback. I don't know what you want to do with the forecast data but once you have the action call there is 5 minutes to create a trigger template to get a sensor were you have all the data in an attribute if you want that...

upsuper commented:
I don't want them to be individual sensors. Also that is a lot of sensors, 48 forecast sensors and 48 actual sensors you mean?

Also as mentioned before, they wouldn't be syncing with the primary sensors because triggers are always async, and because the main sensor may not have its value changed, having to use an interval to trigger makes the gap even larger.

You have not been addressing the issues I was raising before.

upsuper commented:
> it's the path forward for HA in general

I think there are strong cases where I don't care about persistent but I do want atomic updates. It would be good to have a mechanism that allows that to replace attributes.

Requiring people to write extra trigger templates not only make it harder to manage for each individual user, but also has significantly unfavorable compromises comparing to what attributes offer as mentioned above.

gjohansson-ST commented:
@upsuper I don't mind to discuss the approach and the problems that you might have with the proposal on how to continue this PR but may I suggest that you create a discord chat for it or in the community forum (to stop cluttering this PR).

joostlek commented:
> Requiring people to write extra trigger templates not only make it harder to manage for each individual user

But that's a global problem, which requires global solutions, like:

https://github.com/home-assistant/architecture/discussions/1027

upsuper commented:
> but may I suggest that you create a discord chat for it or in the community forum (to stop cluttering this PR).

Thanks for the suggestion. I don't think I have enough energy to initiate larger discussions on this.

If you can take my idea to the team, that would be appreciated, but if you don't think it's worth it, or the team decides action call is the only path forward, I would retreat here and probably seek alternative solutions.

gjohansson-ST commented:
> If you can take my idea to the team, that would be appreciated, but if you don't think it's worth it, or the team decides action call is the only path forward, I would retreat here and probably seek alternative solutions.

Well, that decision has already been made which I tried to inform you quite a few times already. It's not (only) my personal opinion but the path forward for this project.
I think the contribution would have been nice to the project but it's your decision if you don't want too.
